File: 2026-2/sesion-de-clase-08/softprog/softprog-accesodatos/src/main/java/pe/edu/pucp/softprog/dao/impl/AreaDAOImpl.java
package pe.edu.pucp.softprog.dao.impl;

import pe.edu.pucp.softprog.dao.AreaDAO;
import pe.edu.pucp.softprog.db.DBManager;
import pe.edu.pucp.softprog.modelo.rrhh.Area;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class AreaDAOImpl extends RegistroDAOImpl<Area> implements AreaDAO {
    @Override
    public List<Area> findAll() throws SQLException {
        String sql =
                """
                SELECT id, nombre, activo FROM area
                """;
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement cmd = conn.prepareStatement(sql);
            ResultSet rs = cmd.executeQuery()) {

            List<Area> areas = new ArrayList<>();
            while (rs.next()) {
                areas.add(mapear(rs, new Area()));
            }
            return areas;
        }
    }

    @Override
    public Area findById(Integer id) throws SQLException {
        if (id == null) {
            throw new IllegalArgumentException("El id no puede ser nulo");
        }

        String sql =
                """
                SELECT id, nombre, activo FROM area WHERE id = ?
                """;
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement cmd = conn.prepareStatement(sql)) {

            cmd.setInt(1, id);

            try (ResultSet rs = cmd.executeQuery()) {
                return rs.next() ? mapear(rs, new Area()) : null;
            }
        }
    }

    @Override
    public Area findByName(String nombre) throws SQLException {
        if (nombre == null) {
            throw new IllegalArgumentException("El nombre no puede ser nulo");
        }

        String sql =
                """
                SELECT id, nombre, activo FROM area WHERE nombre = ?
                """;
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement cmd = conn.prepareStatement(sql)) {

            cmd.setString(1, nombre);

            try (ResultSet rs = cmd.executeQuery()) {
                return rs.next() ? mapear(rs, new Area()) : null;
            }
        }
    }

    @Override
    public void insert(Area area) throws SQLException {
        if (area == null) {
            throw new IllegalArgumentException("El área no puede ser nula");
        }

        String sql =
                """
                INSERT INTO area (nombre, activo) VALUES (?, ?)
                """;
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement cmd = conn.prepareStatement(sql,
                    PreparedStatement.RETURN_GENERATED_KEYS)) {

            cmd.setString(1, area.getNombre());
            cmd.setBoolean(2, area.isActivo());

            if (cmd.executeUpdate() == 0) {
                throw new SQLException("No se pudo insertar el área");
            }

            try (ResultSet rs = cmd.getGeneratedKeys()) {
                if (rs.next()) {
                    area.setId(rs.getInt(1));
                }
            }
        }
    }

    @Override
    public void update(Area area) throws SQLException {
        if (area == null) {
            throw new IllegalArgumentException("El área no puede ser nula");
        }

        String sql =
                """
                UPDATE area SET nombre = ?, activo = ? WHERE id = ?
                """;
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement cmd = conn.prepareStatement(sql)) {

            cmd.setString(1, area.getNombre());
            cmd.setBoolean(2, area.isActivo());
            cmd.setInt(3, area.getId());

            if (cmd.executeUpdate() == 0) {
                throw new SQLException("No se pudo actualizar el área");
            }
        }
    }

    @Override
    public void delete(Integer id) throws SQLException {
        if (id == null) {
            throw new IllegalArgumentException("El id no puede ser nulo");
        }

        String sql =
                """
                DELETE FROM area WHERE id = ?
                """;
        try (
            Connection conn = DBManager.getInstance().getConnection();
            PreparedStatement cmd = conn.prepareStatement(sql)) {

            cmd.setInt(1, id);

            if (cmd.executeUpdate() == 0) {
                throw new SQLException("No se pudo eliminar el área");
            }
        }
    }

    @Override
    protected Area mapear(ResultSet rs, Area area) throws SQLException {
        super.mapear(rs, area);
        area.setNombre(rs.getString("nombre"));
        return area;
    }
}

File: 2026-2/sesion-de-clase-11/SoftProg/SoftProg.AccessoDatos/DAO/Impl/AreaDAOImpl.cs
using SoftProg.DbManager;
using SoftProg.Modelo.RRHH;
using System.Data;
using System.Data.Common;
using System.Text;

namespace SoftProg.AccessoDatos.DAO.Impl {
    public class AreaDAOImpl : IAreaDAO {
        public List<Area> FindAll() {
            string sql = 
                """
                SELECT id, nombre, activo FROM area
                """;

            using DbConnection conn = DBManager.GetInstance().GetConnection();
            using DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = sql;
            using DbDataReader reader = cmd.ExecuteReader();

            List<Area> areas = [];
            while (reader.Read()) {
                areas.Add(Mapear(reader, new Area()));
            }

            return areas;
        }

        public Area? FindById(int id) {
            string sql =
                """
                SELECT id, nombre, activo 
                FROM area
                WHERE 
                    id = @id
                """;

            using DbConnection conn = DBManager.GetInstance().GetConnection();
            using DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = sql;
            cmd.AgregarParametroEntero("id", id);
            using DbDataReader reader = cmd.ExecuteReader();

            return reader.Read() ? Mapear(reader, new Area()) : null;
        }

        public void Insert(Area area) {
            string sql =
                """
                INSERT INTO area(nombre, activo)
                VALUES (@nombre, @activo);
                SELECT LAST_INSERT_ID();
                """;

            using DbConnection conn = DBManager.GetInstance().GetConnection();
            using DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = sql;
            cmd.AgregarParametroCadena("nombre", area.Nombre);
            cmd.AgregarParametroBoolean("activo", area.IsActivo);
            
            object? resultado = cmd.ExecuteScalar();
            if (resultado == null || resultado == DBNull.Value) {
                throw new Exception("No se pudo insertar el area");
            }

            area.Id = Convert.ToInt32(resultado);
        }

        public void Update(Area area) {
            string sql =
                """
                UPDATE area
                SET 
                    nombre = @nombre, 
                    activo = @activo
                WHERE
                    id = @id
                """;

            using DbConnection conn = DBManager.GetInstance().GetConnection();
            using DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = sql;
            cmd.AgregarParametroCadena("nombre", area.Nombre);
            cmd.AgregarParametroBoolean("activo", area.IsActivo);
            cmd.AgregarParametroEntero("id", area.Id);

            if (cmd.ExecuteNonQuery() == 0) {
                throw new Exception("No pudo acualizar el area");
            }
        }

        public void Delete(int id) {
            string sql =
                """
                DELETE FROM area WHERE id = @id
                """;

            using DbConnection conn = DBManager.GetInstance().GetConnection();
            using DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = sql;
            cmd.AgregarParametroEntero("id", id);

            if (cmd.ExecuteNonQuery() == 0) {
                throw new Exception("No pudo eliminar el area");
            }
        }

        private Area Mapear(DbDataReader reader, Area area) {
            area.Id = reader.GetInt32("id");
            area.Nombre = reader.GetString("nombre");
            area.IsActivo = reader.GetBoolean("activo");
            return area;
        }
    }
}
